pkg/operator/controller/externaldns/deployment.go

```diff
@@ -0,0 +1,261 @@
+var providerStringTable = map[operatorv1alpha1.ExternalDNSProviderType]string{
+	operatorv1alpha1.ProviderTypeAWS:      "aws",
+	operatorv1alpha1.ProviderTypeGCP:      "google",
+	operatorv1alpha1.ProviderTypeAzure:    "azure",
+	operatorv1alpha1.ProviderTypeBlueCat:  "bluecat",
+	operatorv1alpha1.ProviderTypeInfoblox: "infoblox",
+}
+
+// sourceStringTable maps ExternalDNSSourceType values from the
+// ExternalDNS operator API to the source string arugment expected by ExternalDNS.
+var sourceStringTable = map[operatorv1alpha1.ExternalDNSSourceType]string{
+	operatorv1alpha1.SourceTypeRoute:   "openshift-route",
+	operatorv1alpha1.SourceTypeService: "service",
+	// TODO: Add CRD source support
+}
+
+// ensureExternalDNSDeployment ensures that the externalDNS deployment exists.
+// Returns a boolean if the deployment exists, and an error when relevant.
+func (r *reconciler) ensureExternalDNSDeployment(ctx context.Context, namespace, image string, serviceAccount *corev1.ServiceAccount, externalDNS *operatorv1alpha1.ExternalDNS) (bool, *appsv1.Deployment, error) {
+	desired, err := desiredExternalDNSDeployment(namespace, image, serviceAccount, externalDNS)
+	if err != nil {
+		return false, nil, fmt.Errorf("failed to build externalDNS deployment: %w", err)
+	}
+
+	exist, current, err := r.currentExternalDNSDeployment(ctx, namespace, externalDNS)
+	if err != nil {
+		return false, nil, fmt.Errorf("failed to get externalDNS deployment: %w", err)
+	}
+
+	// create the deployment
+	if !exist {
+		if err := r.createExternalDNSDeployment(ctx, desired); err != nil {
+			return false, nil, err
+		}
+		// getting the deployment from API in case a third party updated it after the creation
+		return r.currentExternalDNSDeployment(ctx, namespace, externalDNS)
+	}
+
+	// update the deployment
+	if _, err := r.updateExternalDNSDeployment(ctx, current, desired); err != nil {
+		return true, current, err
+	}
+
+	// getting the deployment from API in case a third party updated it after the update
+	return r.currentExternalDNSDeployment(ctx, namespace, externalDNS)
+}
+
+// currentExternalDNSDeployment gets the current externalDNS deployment resource.
+func (r *reconciler) currentExternalDNSDeployment(ctx context.Context, namespace string, externalDNS *operatorv1alpha1.ExternalDNS) (bool, *appsv1.Deployment, error) {
+	depl := &appsv1.Deployment{}
+	if err := r.client.Get(ctx, types.NamespacedName{Namespace: namespace, Name: externalDNS.Name}, depl); err != nil {
+		if errors.IsNotFound(err) {
+			return false, nil, nil
+		}
+		return false, nil, err
+	}
+	return true, depl, nil
+}
+
+// desiredExternalDNSDeployment returns the desired deployment resource.
+func desiredExternalDNSDeployment(namespace, image string, serviceAccount *corev1.ServiceAccount, externalDNS *operatorv1alpha1.ExternalDNS) (*appsv1.Deployment, error) {
+	replicas := int32(1)
+
+	matchLbl := map[string]string{
+		"app.kubernetes.io/name":     "external-dns",
```

Comment:
Would be nice to have all these strings declared as consts

Reply:
I put a constant for externalDNS to `names.go` file in `pkg/operator/controller/` package.
For the label names, I think it's fine to have them defined here. I don't see a big advantage of putting them as `const` at the beginning of the file, I see them more useful to be defined closer to the instantiation of `appsv1.Deployment`.

Reply:
Just noticed that in my next PR I used the same value for the toleration's key as the node selector's label.
Will move this value to const and to keep it consistent will do this for all.